checker/src/main/java/org/checkerframework/checker/optional/OptionalAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,112 @@
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.TreeUtils;
+
+/** OptionalAnnotatedTypeFactory for the Optional Checker. */
+public class OptionalAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+  /** The element for java.util.Optional.map(). */
+  private final ExecutableElement optionalMap;
+
+  /** The @{@link Present} annotation. */
+  protected final AnnotationMirror PRESENT = AnnotationBuilder.fromClass(elements, Present.class);
+
+  /**
+   * Creates an OptionalAnnotatedTypeFactory.
+   *
+   * @param checker a checker
+   */
+  public OptionalAnnotatedTypeFactory(BaseTypeChecker checker) {
+    super(checker);
+    postInit();
+    optionalMap = TreeUtils.getMethodOrNull("java.util.Optional", "map", 1, getProcessingEnv());
+  }
+
+  @Override
+  public AnnotatedTypeMirror getAnnotatedType(Tree tree) {
+    AnnotatedTypeMirror result = super.getAnnotatedType(tree);
+    optionalMapNonNull(tree, result);
+    return result;
+  }
+
+  /**
+   * If {@code tree} is a call to {@link java.util.Optional#map(Function)} whose argument is a
+   * method reference, and if the option {@code "optionalMapAssumeNonNull"} is passed, then this
+   * method adds {@code @Present} to {@code type} if the follow is true:
+   *
+   * <p>The type of the receiver to {@link java.util.Optional#map(Function)} is {@code @Present}
+   *
+   * <p>And the return type of the function type represented by the method reference that is an
+   * argument to {@code map}, is not explicitly annotated with a {@code Nullable} annotation.
+   *
+   * @param tree a tree
+   * @param type a type that is side-effected by this method
+   */
+  private void optionalMapNonNull(Tree tree, AnnotatedTypeMirror type) {
+    if (!TreeUtils.isMethodInvocation(tree, optionalMap, processingEnv)) {
+      return;
+    }
+    MethodInvocationTree mapTree = (MethodInvocationTree) tree;
+    ExpressionTree argTree = mapTree.getArguments().get(0);
+    if (argTree.getKind() == Kind.MEMBER_REFERENCE) {
+      MemberReferenceTree memberReferenceTree = (MemberReferenceTree) argTree;
+      AnnotatedTypeMirror optType = getReceiverType(mapTree);
+      if (optType == null || !optType.hasEffectiveAnnotation(Present.class)) {
+        return;
+      }
+      if (TreeUtils.MemberReferenceKind.getMemberReferenceKind(memberReferenceTree)
+          .isConstructorReference()) {
+        // Constructors are always non-null.
+        type.replaceAnnotation(PRESENT);
+        return;
+      }
+      if (!checker.hasOption("optionalMapAssumeNonNull")) {
+        return;
+      }
+      ExecutableElement memberReferenceFuncType = TreeUtils.elementFromUse(memberReferenceTree);
+      if (!containsNullable(memberReferenceFuncType.getAnnotationMirrors())
+          && !containsNullable(memberReferenceFuncType.getReturnType().getAnnotationMirrors())) {
+        type.replaceAnnotation(PRESENT);
+      }
```

Comment:
Could you abstract this code out to a method that returns a boolean?  Then this code will be
```if (...) { type.replaceAnnotation(PRESENT); }```
I think that will make this code clearer, will provide a good place for special-case code like constuctors, and will enable us to simplify the Javadoc of `optionalMapNonNull`, which is currently quite a mouthful (and the part about "if the option is passed" appears in the wrong part of the Javadoc).